File: master/internal/geo/geo.go
// Package geo resolves approximate human-readable regions from IP addresses
// using the free ip-api.com service. It is a soft dependency: any failure
// returns "" so callers degrade gracefully.
package geo

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// Region returns a short region label (e.g. "HK · 香港 · 香港岛") for an IP.
// Private/loopback addresses and any lookup failure yield "".
func Region(ip string) string {
	ip = strings.TrimSpace(ip)
	if ip == "" || ip == "127.0.0.1" || ip == "::1" || strings.EqualFold(ip, "localhost") {
		return ""
	}
	u := fmt.Sprintf("http://ip-api.com/json/%s?fields=status,regionName,city,countryCode&lang=zh-CN",
		url.QueryEscape(ip))
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(u)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()

	var g struct {
		Status      string `json:"status"`
		CountryCode string `json:"countryCode"`
		RegionName  string `json:"regionName"`
		City        string `json:"city"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&g); err != nil || g.Status != "success" {
		return ""
	}
	var parts []string
	if g.CountryCode != "" {
		parts = append(parts, g.CountryCode)
	}
	if g.RegionName != "" && g.RegionName != g.CountryCode {
		parts = append(parts, g.RegionName)
	}
	if g.City != "" && g.City != g.RegionName {
		parts = append(parts, g.City)
	}
	return strings.Join(parts, " · ")
}

File: web/src/lib/geo.ts
// IP 探测辅助：用公共服务自动获取本机公网 IPv4 / IPv6 与地区。
// 全部为软依赖——任一失败返回空字符串，调用方优雅降级。

export interface IpInfo {
  ipv4: string
  ipv6: string
  region: string
}

export async function detectIpInfo(): Promise<IpInfo> {
  const [v4, v6, geo] = await Promise.allSettled([
    fetch('https://api.ipify.org?format=json').then((r) => r.json() as Promise<{ ip?: string }>),
    fetch('https://api6.ipify.org?format=json').then((r) => r.json() as Promise<{ ip?: string }>),
    fetch(
      'https://ip-api.com/json/?fields=status,message,query,city,regionName,countryCode&lang=zh-CN',
    ).then((r) => r.json() as Promise<{ status?: string; countryCode?: string; regionName?: string; city?: string }>),
  ])

  const ipv4 = v4.status === 'fulfilled' ? (v4.value.ip ?? '') : ''
  const ipv6 = v6.status === 'fulfilled' ? (v6.value.ip ?? '') : ''

  let region = ''
  if (geo.status === 'fulfilled' && geo.value.status === 'success') {
    const parts = [geo.value.countryCode, geo.value.regionName, geo.value.city].filter(
      (p, i, arr) => p && arr.indexOf(p) === i,
    )
    region = parts.join(' · ')
  }

  return { ipv4, ipv6, region }
}
